### PR TITLE
HHH-14251 Invalid SQL for @Embedded UPDATE

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-14251

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/test/hql/UpdateEntityWithEmbeddedTest.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import org.hibernate.testing.RequiresDialect;
+import org.hibernate.testing.TestForIssue;
+import org.hibernate.testing.junit4.BaseCoreFunctionalTestCase;
+import org.junit.Test;
+
+/**
+ * @author Andrea Boriero
+ */
+@TestForIssue(jiraKey = "HHH-14251")
+@RequiresDialect(PostgreSQL81Dialect.class)
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/test/hql/UpdateEntityWithEmbeddedTest.java:25]: Is the issue dialect specific? Saw the code changes in `DotNode` are generic.

Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/test/hql/UpdateEntityWithEmbeddedTest.java:25]: Different behaviour based on `Dialect#supportsRowValueConstructorSyntax()` , I'm working on a fix for Dialects not supporting it.

File: hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/BinaryLogicOperatorNode.java [hunk 2]
```diff
@@ -164,14 +167,20 @@ protected void translate(
 
 		for ( int i = valueElements - 1; i > 0; i-- ) {
 			if ( i == 1 ) {
-				AST op1 = getASTFactory().create( comparisonType, comparisonText );
+				final AST op1;
+				if ( isInsideSetClause() ) {
+					op1 = container;
+				}else {
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/BinaryLogicOperatorNode.java:173]: super minor. we usually adopt the following code template
```
if () {
}
else {
}
```

File: hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/BinaryLogicOperatorNode.java [hunk 3]
```diff
@@ -206,6 +219,10 @@ protected void translate(
 		}
 	}
 
+	private boolean isInsideSetClause() {
+		return getWalker().getCurrentClauseType() == HqlSqlTokenTypes.SET;
+	}
+
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/BinaryLogicOperatorNode.java:225]: feel free to Ignore my comment for I am not sure what I am saying, :). Usually `INSERT` and `SET` follow the same or similar syntax. Do your code changes fix the insert scenario, btw? If not, should we include `INSERT` above?

File: hibernate-core/src/test/java/org/hibernate/test/hql/UpdateEntityWithEmbeddedTest.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+		inTransaction(
+				session -> {
+					session.createQuery(
+							"UPDATE Company c SET c.logo.fileName = :filename, c.logo.fileExtension = :fileExtension" )
+							.setParameter( "filename", "logo2" )
+							.setParameter( "fileExtension", "png" )
+							.executeUpdate();
+				}
+		);
+	}
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/test/hql/UpdateEntityWithEmbeddedTest.java:74]: Could we include 'insert' testing cases above?

File: hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/BinaryLogicOperatorNode.java (hunk 2 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/BinaryLogicOperatorNode.java]: Fixed, 
Thanks @NathanQingyangXu

File: hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/BinaryLogicOperatorNode.java (hunk 3 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/BinaryLogicOperatorNode.java:226]: HQL INSERT does not follow the same syntax used for the SET https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#_hql_syntax_for_insert.

Comment (NathanQingyangXu) [reply hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/BinaryLogicOperatorNode.java:226]: I meant `insert into ... values (...)`

File: hibernate-core/src/test/java/org/hibernate/test/hql/UpdateEntityWithEmbeddedTest.java (hunk 1 above)
Comment (quaff) [reply hibernate-core/src/test/java/org/hibernate/test/hql/UpdateEntityWithEmbeddedTest.java]: > Different behaviour based on `Dialect#supportsRowValueConstructorSyntax()` , I'm working on a fix for Dialects not supporting it.

https://github.com/hibernate/hibernate-orm/blob/41161f9fa9e1d760995c462c78cbd19ef8e7b082/hibernate-core/src/main/java/org/hibernate/dialect/MySQLDialect.java#L456
`MySQLDialect#supportsRowValueConstructorSyntax()` return true, but row value constructor is not supported in set clause, only supported in where clause.

Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/test/hql/UpdateEntityWithEmbeddedTest.java]: Thanks @quaff, I'm working on a possible solution.

**Comment by Sanne** [comment]
What's the status of this one?

**Comment by dreab8** [comment]
I committed a change that should fix the issue raised by @quaff ( MySQL does not support row value constructor in set clause, only in where clause).

About Nathan's comment, for what I know hql does not support `insert into ... values (...)` syntax (https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#_hql_syntax_for_insert) but may be I'm wrong or I have not understood his comment

**Comment by NathanQingyangXu** [comment]
Well, I am not sure about my comment. I think hql does support that  if
you check the ANTLR grammar, but feel free to ignore my comment.

**Comment by dreab8** [comment]
Hi @NathanQingyangXu ,

the antlr grammar for insert is
```
insertStatement
	 INSERT^ intoClause selectStatement
	;
```
and the documentation (https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#_hql_syntax_for_insert states
_Only the INSERT INTO …​ SELECT …​ form is supported. You cannot specify explicit values to insert._

but may be I'm just misunderstanding your comment

**Comment by NathanQingyangXu** [comment]
> Hi @NathanQingyangXu ,
> 
> the antlr grammar for insert is
> 
> ```
> insertStatement
> 	 INSERT^ intoClause selectStatement
> 	;
> ```
> 
> and the documentation (https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#_hql_syntax_for_insert states
> _Only the INSERT INTO …​ SELECT …​ form is supported. You cannot specify explicit values to insert._
> 
> but may be I'm just misunderstanding your comment

You understood my comment correctly. I saw `insert ... values ...` is only supported in v6, not in v5. Thanks for the following up.

**Comment by beikov** [comment]
I know for a fact that Hibernate < 6 does not support something like `insert into Entity(embeddable.property1, embeddable.property2, ...) select ...` but handling that in this PR is probably an overkill.

**Review by Sanne — APPROVED** [review]

**Comment by Sanne** [comment]
Thanks for all feedback! I'll merge it.

**Pull request closed without merging**